pkg/gadgets/trace/tcpdrop/tracer/tracer.go

```diff
@@ -0,0 +1,246 @@
+	}
+	t.reader = reader
+
+	return nil
+}
+
+func (t *Tracer) run() {
+	for {
+		record, err := t.reader.Read()
+		if err != nil {
+			if errors.Is(err, perf.ErrClosed) {
+				// nothing to do, we're done
+				return
+			}
+
+			msg := fmt.Sprintf("Error reading perf ring buffer: %s", err)
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
+
+		if record.LostSamples > 0 {
+			msg := fmt.Sprintf("lost %d samples", record.LostSamples)
+			t.eventCallback(types.Base(eventtypes.Warn(msg)))
+			continue
+		}
+
+		bpfEvent := (*tcpdropEvent)(unsafe.Pointer(&record.RawSample[0]))
+
+		reason := lookupDropReason(int(bpfEvent.Reason))
+
+		event := types.Event{
+			Event: eventtypes.Event{
+				Type:      eventtypes.NORMAL,
+				Timestamp: gadgets.WallTimeFromBootTime(bpfEvent.Timestamp),
+			},
+			WithMountNsID: eventtypes.WithMountNsID{MountNsID: bpfEvent.MntnsId},
+			Pid:           bpfEvent.Pid,
+			Comm:          gadgets.FromCString(bpfEvent.Comm[:]),
+			Dport:         gadgets.Htons(bpfEvent.Dport),
+			Sport:         gadgets.Htons(bpfEvent.Sport),
+			State:         tcpbits.TCPState(bpfEvent.State),
+			Tcpflags:      tcpbits.TCPFlags(bpfEvent.Tcpflags),
+			Reason:        reason,
+		}
+
+		if bpfEvent.Af == unix.AF_INET {
+			event.IPVersion = 4
+		} else if bpfEvent.Af == unix.AF_INET6 {
+			event.IPVersion = 6
+		}
+
+		event.Saddr = gadgets.IPStringFromBytes(bpfEvent.Saddr, event.IPVersion)
+		event.Daddr = gadgets.IPStringFromBytes(bpfEvent.Daddr, event.IPVersion)
+
+		if t.enricher != nil {
+			t.enricher.EnrichByMntNs(&event.CommonData, event.MountNsID)
+		}
+
+		t.eventCallback(&event)
+	}
+}
+
+var (
+	btfDetectDropReason sync.Once
+	dropReasons         = map[int]string{}
+	dropReasonsErr      error
+)
+
+func lookupDropReason(reason int) string {
+	btfDetectDropReason.Do(func() {
+		btfSpec, err := btf.LoadKernelSpec()
+		if err != nil {
+			dropReasonsErr = err
+			return
+		}
+
+		enum := &btf.Enum{}
+		err = btfSpec.TypeByName("skb_drop_reason", &enum)
+		if err != nil {
+			dropReasonsErr = err
+			return
+		}
+		for _, v := range enum.Values {
+			dropReasons[int(v.Value)] = v.Name
+		}
+	})
+
+	if dropReasonsErr != nil {
+		return dropReasonsErr.Error()
+	}
+	if ret, ok := dropReasons[reason]; ok {
+		ret = strings.TrimPrefix(ret, "SKB_DROP_REASON_")
+		ret = strings.TrimPrefix(ret, "SKB_")
+		return ret
+	} else {
+		return fmt.Sprintf("UNKNOWN#%d", reason)
+	}
+}
+
+// --- Registry changes
+
+func (t *Tracer) Run(gadgetCtx gadgets.GadgetContext) error {
+	defer t.close()
+	if err := t.install(); err != nil {
+		return fmt.Errorf("installing tracer: %w", err)
+	}
+
+	go t.run()
+	gadgetcontext.WaitForTimeoutOrDone(gadgetCtx)
+
+	return nil
+}
+
+func (t *Tracer) SetMountNsMap(mountnsMap *ebpf.Map) {
+	// tcpdrop does not filter by container but we need to implement
+	// MountNsMapSetter, otherwise LocalManager.CanOperateOn() returns false
+	// and events won't be enriched.
```

Comment:
Why can't it filter by container? Is there a technical limitation of just something that is not implemented?

Reply:
See discussion in https://github.com/inspektor-gadget/inspektor-gadget/pull/1469#discussion_r1152105218